File: typescript-examples/src/module1/testDoubles/mock/__mocks__/MockLogger.ts
import {Logger} from "../Logger.ts";

class AssertionError extends Error {
}

export class MockLogger implements Logger {
    private readonly loggedMessages: string[] = [];
    private readonly expectedMessages: string[] = [];

    public log(message: string): void {
        this.loggedMessages.push(message);
    }

    public expect(message: string) {
        this.expectedMessages.push(message);
    }

    public verify(): void {
        this.assertExpectsSize();

        this.assertLoggedMessages();
    }

    private assertExpectsSize() {
        if (this.loggedMessages.length !== this.expectedMessages.length) {
            this.throwMessageSizeMismatchError(this.expectedMessages.length, this.loggedMessages.length);
        }
    }

    private throwMessageSizeMismatchError(expectedSize: number, givenSize: number) {
        throw new AssertionError(`Expected messages size did not match logged messages size: Expected ${expectedSize} but given ${givenSize}`)
    }

    private throwMessageMismatchError(expectedMessage: string, givenMessage: string) {
        throw new AssertionError(`Expected messages did not match logged messages: Expected ${expectedMessage} but given ${givenMessage}`)
    }

    private assertLoggedMessages() {
        for (let messageIndex = 0; messageIndex < this.loggedMessages.length; messageIndex++) {
            if (this.loggedMessages[messageIndex] !== this.expectedMessages[messageIndex]) {
                this.throwMessageMismatchError(this.expectedMessages[messageIndex], this.loggedMessages[messageIndex])
            }
        }
    }
}

File: module-1-content/src/test/java/com/codurance/module1/TestDoublesMockito/Mock/MockLogger.java
package com.codurance.module1.TestDoublesMockito.Mock;

import com.codurance.module1.TestDoubles.Mock.Logger;

import java.util.ArrayList;
import java.util.List;

public class MockLogger implements Logger {
    private final List<String> loggedMessages = new ArrayList<>();
    private final List<String> expectedMessages = new ArrayList<>();

    @Override
    public void log(String message) {
        loggedMessages.add(message);
    }

    // Method to set expected messages
    public void expect(String message) {
        expectedMessages.add(message);
    }

    // Verification method
    public void verify() {
        assertExpectsSize();

        assertLoggedMessages();
    }

    private void assertLoggedMessages() {
        for (int messageIndex = 0; messageIndex < loggedMessages.size(); messageIndex++) {
            if (loggedMessages.get(messageIndex).equals(expectedMessages.get(messageIndex))) {
                continue;
            }

            throwAssertionException(expectedMessages.get(messageIndex), loggedMessages.get(messageIndex));
        }
    }

    private void assertExpectsSize() {
        if (loggedMessages.size() != expectedMessages.size()) {
            throwAssertionException(expectedMessages.size(), loggedMessages.size());
        }
    }

    private void throwAssertionException(int expectedSize, int givenSize) {
        throw new AssertionError(String.format(
            "Expected messages size did not match logged messages size: Expected \"%d\" bu \"%d\" given.",
            expectedSize,
            givenSize
        ));
    }

    private void throwAssertionException(String expectedMessage, String givenMessage) {
        throw new AssertionError(String.format(
                "Expected messages did not match logged messages: Expected \"%s\" but \"%s\" given.",
            expectedMessage,
            givenMessage
        ));
    }
}
